xla/service/gpu/ir_emission_utils.cc

```diff
@@ -740,6 +739,90 @@ bool IsIntermediate(const HloInstruction* instr, int allowed_operand_count) {
                                       instr->shape(), instr->dimensions()))));
 }
 
+bool IsReduceIntermediate(const HloInstruction* instr) {
+  if (instr->operand_count() > 1 || instr->user_count() > 1) {
+    return false;
+  }
+
+  // Only support elementwise ops that don't introduce additional compute.
+  // More benchmarking and better cost model are needed to enable this for
+  // more compute ops.
+  switch (instr->opcode()) {
+    case HloOpcode::kBitcast:
+    case HloOpcode::kBitcastConvert:
+    case HloOpcode::kConvert:
+      return true;
+    case HloOpcode::kReshape:
+      return ShapeUtil::ReshapeIsBitcast(instr->operand(0)->shape(),
+                                         instr->shape());
+    default:
+      return false;
+  }
+}
+
+// Recursive helper for GetFusionRoots below.
+static void GetFusionRootsRec(HloInstruction* root,
```

Comment:
Can this be moved back to gpu_fusible?

Reply:
i will this in FindRealReductionHero, if we move this to gpu_fusible, then i need to include the header which would cause a circular dependency in bazel files..